activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -1514,6 +1572,35 @@ def build_select(arel)
         end
       end
 
+      def build_with(arel)
+        return if with_values.empty?
+
+        with_statements = with_values.map do |with_value|
+          case with_value
+          when Arel::Nodes::As, Arel::Nodes::TableAlias then with_value
+          when Hash then build_with_value_from_hash(with_value)
+          else
+            raise ArgumentError, "Unsupported argument type: #{with_value} #{with_value.class}"
+          end
+        end
+
+        arel.with(with_statements)
+      end
+
+      def build_with_value_from_hash(hash)
+        hash.map do |name, value|
+          expression =
+            case value
+            when Arel::Nodes::SqlLiteral then Arel.sql("(#{value})")
+            when ActiveRecord::Relation then value.arel
+            when Arel::SelectManager then value
+            else
+              raise ArgumentError, "Unsupported argument type: `#{value}` #{value.class}"
+            end
+          Arel::Nodes::TableAlias.new(expression, Arel.sql(name.to_s))
```

Comment:
I don't know the specific node this should be off the top of my head, but it definitely should not inject the hash key as raw SQL; `name` should end up table/column quoted.

Reply:
Thanks for catching those! Both fixed in https://github.com/rails/rails/pull/37944/commits/5b8676c3c230a3d3c2deeba07cce0949636e0132